[C++-Interop] Import `const &T` function parameters as @in_guaranteed

plotfi opened the pull request:
When passing value types as a const-ref to a C++ API, ensure that the caller is tasked with handling the lifetime of the instance passed in.

@hyp @zoecarver @drodriguez

plotfi committed:
[C++-Interop] Import `const &T` function parameters as @in_guaranteed

When passing value types as a const-ref to a C++ API, ensure that the
caller is tasked with handling the lifetime of the instance passed in.

drodriguez commented:
@swift-ci please smoke test

drodriguez commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -42,36 +42,36 @@ func testFunction() {
 // CHECK-SAME: : $@convention(method) (OptionsStruct, @thick OptionsConsumerObjC.Type) -> @owned OptionsConsumerObjC
 
 // COM: FIXME: should it be @in_guaranteed OptionStruct? https://github.com/apple/swift/issues/60601
```
Remove? Same for the other cases.

drodriguez commented:
Not for this PR necessarily, but we are still missing the one in the allocator/initializer case 😞

plotfi commented:
> Not for this PR necessarily, but we are still missing the one in the allocator/initializer case 😞

Added `test/Interop/Cxx/reference/reference-silgen-cxx-objc-ctors+init.swift` which shows that the ObjC init is Ok, but the C++ ctor is not getting any sort of `@in` or `@in_guranteed` so I added a FIXME to track this. Will file an issue too.

drodriguez commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -36,42 +36,34 @@ func testFunction() {
 // RUN: %target-swift-frontend -c -enable-experimental-cxx-interop -enable-objc-interop -I %S/Inputs %s -emit-silgen -o - | %FileCheck %s
 
-// COM: FIXME: should it be @in_guaranteed OptionsStruct? https://github.com/apple/swift/issues/60601
 // CHECK: [[FN1:%[0-9]+]] = function_ref @$sSo19OptionsConsumerObjCC7optionsABSo0A6StructV_tcfC : $@convention(method) (OptionsStruct, @thick OptionsConsumerObjC.Type) -> @owned OptionsConsumerObjC
```
This one is the one that I was confused. IIUC `$sSo19OptionsConsumerObjCC7optionsABSo0A6StructV_tcfC` demangles to `__C.OptionsConsumerObjC.__allocating_init(options: __C.OptionsStruct) -> __C.OptionsConsumerObjC` which seems to be an Obj-C initializer, and it does not have `@in` nor `@in_guaranteed`. I don't know if this is required, or in this case it is  fine to pass a copy.

But this one is different than in your new test, which shows a non-mangled name. I wonder if the designated initializer bit changes something.

drodriguez commented on test/reference-silgen_out.swift:
```diff
@@ -0,0 +1,182 @@
+sil_stage canonical
```
Not sure you wanted to submit this file to the upstream.

plotfi replied:
Ah thank you! This was a mistake.

bulbazord commented:
@swift-ci please smoke test

drodriguez commented:
@swift-ci please smoke test

plotfi commented:
@hyp Do these changes look ok to you?

zoecarver commented on lib/SILGen/SILGenBridging.cpp:
```diff
@@ -2225,7 +2225,8 @@ void SILGenFunction::emitForeignToNativeThunk(SILDeclRef thunk) {
                                                 foreignFormalType,
                                                 foreignLoweredTy);
         // Handle C pointer arguments imported as indirect `self` arguments.
```
This comment is no longer correct. Or maybe this isn't the right place for this logic?

plotfi replied:
Agreed. I am not even sure what the comment is trying to clarify here.

zoecarver commented on test/Interop/Cxx/reference/Inputs/reference-silgen-cxx-objc-ctors+init.h:
```diff
@@ -0,0 +1,18 @@
+
+@interface ObjCSwiftBridgeT : NSObject
+- (instancetype)init  __attribute__((unavailable));
+- (instancetype)initWithEmbedded:(const CxxEmbeddedT &)embedded __attribute__((objc_designated_initializer));
```
Is all of this needed? We can't just take a `PrimitiveWrapperT` here?

plotfi replied:
I think I was trying to see if adding an extra layer of wrapper effects if it is passed by value or as a @in*

plotfi replied:
I remember now, for the `FIXME`

I wanted to show that the call to  `_ZN17IntWrapperWrapperC1ERK10IntWrapper` is a

` function_ref @_ZN17IntWrapperWrapperC1ERK10IntWrapper : $@convention(c) (IntWrapper)`

and not a 

` function_ref @_ZN17IntWrapperWrapperC1ERK10IntWrapper : $@convention(c) (@in_guaranteed  IntWrapper)`

plotfi commented:
Addressed your comments @zoecarver

hyp reviewed (approved):
LGTM
